### PR TITLE
Ability to create a joined table without downloading tables assets

### PR DESCRIPTION
**Pull request by tssweeney**
In this PR, i updated the JoinedTable constructor to allow the user to pass in results from Artifact#get_path(TABLE_NAME). The motivation is to enable JoinedTable construction without actually downloading the table data itself. This involves a few changes:

1. Allow the user to `Artifact.get_path(NAME)` on objects which were added with `Artifact.add(obj, NAME)`. This returns an artifact entry object which can be passed to the table. Moreover, this maintains our current API schema where `get_path` returns a reference and `get` instantiates the object itself.
2. All the JoinedTable constructor to accept this kind of object
3. Update the `to_json` method to handle this object properly
4. Associated integration tests.

### CODE REVIEW

File: standalone_tests/artifact_object_reference_test.py [hunk 1]
```diff
@@ -508,6 +508,49 @@ def test_joined_table_referential():
...
+    with wandb.init(project=PROJECT_NAME) as run:
+        tables = wandb.Artifact("tables", "database")
+        tables.add(src_table_1, "src_table_1")
+        tables.add(src_table_2, "src_table_2")
+
+        # Should be able to add by name directly
+        jt = wandb.JoinedTable("src_table_1.table.json", "src_table_2.table.json", "id")
+        tables.add(jt, "jt")
+
+        # Make sure it errors when you are not referencing the correct table names
```
Comment (shawnlewis) [inline standalone_tests/artifact_object_reference_test.py:527]: Thank you for testing error cases :)

**Review by shawnlewis — APPROVED** [review]

**Review by vanpelt — APPROVED** [review]
Nice, for some of this lower level logic it probably makes more sense to add regular unit tests.  It's not clear to me if these standalone_tests are actually getting run in CI or systematically in the regression suite...

**Comment by tssweeney** [comment]
> Nice, for some of this lower level logic it probably makes more sense to add regular unit tests. It's not clear to me if these standalone_tests are actually getting run in CI or systematically in the regression suite...

That is a great point, I should have brought some of this into the unit tests. I try to do that as much as i can (whenever there is not need to test the end to end logic), but missed it here. 

FWIW - this next release (and all futures) will have the artifact standalone included in regression.

File: standalone_tests/artifact_object_reference_test.py (hunk 1 above)
Comment (tssweeney) [reply standalone_tests/artifact_object_reference_test.py:527]: 👍

**Pull request merged**